### PR TITLE
Add other joins and test cases

### PR DESCRIPTION
**Pull request by srfc**
## Proposed Changes

  - Add `innerJoin`, `outerJoin`, `leftJoin`, `rightJoin` that work using a comparator function instead of column equality
  - Add test cases 

## Related

  - this PR

### CODE REVIEW

File: adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/FutureRelationTest.scala [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+      .build()
+
+    "empty" should {
+      val futureEmpty: Future[Seq[Record]] = Future {
+        Thread.sleep(200)
+        Seq.empty
+      }
+      val emptyRelation = FutureRelation.fromRecordSeq(futureEmpty)
+
+      "return an empty result set for any where or whereAll query" in {
```
Comment (CodeLionX) [inline adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/FutureRelationTest.scala]: we already checked this behavior in `TransientRelation`, so you don't have to copy it here for our small project ;)
But as you already did, we keep it there.

File: adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/FutureRelationTest.scala [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+                .withCellContent(colCustomerId)(7)
+                .withCellContent(colFullname)("Omari Wesson")
+                .withCellContent(colCountry)("USA")
+                .build()
+            ))
+        }
+      }
+    }
+  }
+}
```
Comment (CodeLionX) [inline adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/FutureRelationTest.scala]: I personally think, we should move all of those to `TransientRelationTest` as this logic is implemented there. This test can then test the parallel execution of those operations and additional operators that will follow.

File: adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/FutureRelationTest.scala (hunk 1 above)
Comment (srfc) [reply adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/FutureRelationTest.scala]: You're right in that this is a little redundant. I wanted to make sure that it still works wrapped in a Future.

File: adbms/src/main/scala/de/up/hpi/informationsystems/adbms/definition/Record.scala [hunk 3]
```diff
@@ -3,7 +3,8 @@ package de.up.hpi.informationsystems.adbms.definition
 import java.util.Objects
 
 import scala.collection.{GenTraversableOnce, MapLike, mutable}
-import scala.util.Try
+import scala.util.{Success, Try}
```
[the remark this answers is not in the document]
Comment (srfc) [reply adbms/src/main/scala/de/up/hpi/informationsystems/adbms/definition/Record.scala]: You're right, this is an artefact just like `AsyncWordSpec` in FutureRelationTest (`Async*Spec` didn't play nice with our `awaitable` `FutureRelation` wrapper - it wants Futures to which you can map tests; only those get executed).

File: adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/FutureRelationTest.scala (hunk 2 above)
Comment (srfc) [reply adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/FutureRelationTest.scala]: But we already have these tests for `TransientRelation` in there. How would you tests parallel execution of them there?

Comment (CodeLionX) [reply adbms/src/test/scala/de/up/hpi/informationsystems/adbms/definition/FutureRelationTest.scala]: Don't know yet :laughing:

**Pull request merged**